src/mm/checkpoints.c

```diff
@@ -41,79 +41,198 @@
-
-		return true;
+		clean_bitmap(blocks, m_area->dirty_bitmap);
+		
+		return true; /// do not log this area skip to the next one
 	}
 
-	return false;
+	return false; /// take a snapshot (full or incremental)
 }
 
-void log_all_marea_chunks(malloc_area *m_area, void **ptr, int bitmap_blocks) {
 
-	size_t chunk_size;
+
+void copy_chunks_in_marea(malloc_area *m_area, unsigned int *bitmap, int blocks, void **ptr, size_t chunk_size) {
+
 	int j, k, idx;
+	// Copy only the in-use chunks
+	for(j = 0; j < blocks; j++){
+
+		// Check the allocation bitmap on a per-block basis, to enhance scan speed
+		if(bitmap[j] == 0) {
+			// Empty (no-chunks-allocated) block: skip to the next
+			continue;
+
+		} else {
+			// At least one chunk is in-use: per-bit scan of the block is required
+			for(k = 0; k < NUM_CHUNKS_PER_BLOCK; k++){
+
+				if(CHECK_BIT_AT(bitmap[j], k)){
+
+					idx = j * NUM_CHUNKS_PER_BLOCK + k;
+					memcpy(*ptr, (void*)((char*)m_area->area + (idx * chunk_size)), chunk_size);
+					*ptr = (void*)((char*) *ptr + chunk_size);
+
+				}
+			}
+		}
+	}
+}
+
+
+void log_all_marea_chunks(malloc_area *m_area, void **ptr, int bitmap_blocks, bool is_incremental) {
+
+	size_t chunk_size;
 
 	chunk_size = m_area->chunk_size;
 	RESET_BIT_AT(chunk_size, 0);	// ckpt Mode bit
 	RESET_BIT_AT(chunk_size, 1);	// Lock bit
 
 	// Check whether the area should be completely copied (not on a per-chunk basis)
 	// using a threshold-based heuristic
-	if(CHECK_LOG_MODE_BIT(m_area)){
+	if(!is_incremental && CHECK_LOG_MODE_BIT(m_area)){
 
 		// If the malloc_area is almost (over a threshold) full, copy it entirely
 		memcpy(*ptr, m_area->area, m_area->num_chunks * chunk_size);
 		*ptr = (void*)((char*) *ptr + m_area->num_chunks * chunk_size);
 
 	} else {
-		// Copy only the allocated chunks
-		for(j = 0; j < bitmap_blocks; j++){
 
-			// Check the allocation bitmap on a per-block basis, to enhance scan speed
-			if(m_area->use_bitmap[j] == 0) {
-				// Empty (no-chunks-allocated) block: skip to the next
-				continue;
+		/// copy the use bitmap regardless of the type of log
+		memcpy(*ptr, m_area->use_bitmap, bitmap_blocks * BLOCK_SIZE);
+		*ptr = (void*)((char*) *ptr + bitmap_blocks * BLOCK_SIZE);
+		
+		/// iss = 0 : original code 
+		/// iss = 1 AND mprotect = 1 → depends on the type of log (full/incremental)
+		///							   full: only copy use bitmap
+		///							   incremental: only copy dirty bitmap
+		/// iss = 1 AND mprotect = 1 → copy both use_bitmap and dirty_bitmap
 
-			} else {
-				// At least one chunk is allocated: per-bit scan of the block is required
-				for(k = 0; k < NUM_CHUNKS_PER_BLOCK; k++){
+		/// copy use bitmap when iss is not enabled
+		/// when iss is enabled only copy the use bitmap if the log is not incremental or mprotect is enabled
+		if ( !(pdes_config.checkpointing == INCREMENTAL_STATE_SAVING) || ( pdes_config.checkpointing == INCREMENTAL_STATE_SAVING && (!is_incremental || pdes_config.iss_enabled_mprotection == MPROTECT) ) ) {
```

Comment:
the case is incremental, MPROTECT and INCREMENTAL_STATE_SAVING is incorrect
```suggestion
		if ( !(pdes_config.checkpointing == INCREMENTAL_STATE_SAVING) || ( pdes_config.checkpointing == INCREMENTAL_STATE_SAVING && (!is_incremental) ) ) {
```